### DISCUSSION TITLE
How to create a framework with five categories and their respective terms ?

### DISCUSSION THREAD

**Discussion by devendra-tarento** [Sunbird-Ed/Community#363] [General Discussion]
We have a requirement of creating a Framework with below categories:

1- Function/position
2- Role
3- Activity
4- Competency
5- Competency Level

I have observed that to create a framework with only 4 categories are getting created by mapping their categories against below 4 sunbird master categories:
1- Board
2- Medium
3- Grade Level
4- Subject

Now We have below queries:
1- Do we have any other master category to which our fifth category(Competency Level) can be mapped ? 
2- If there is a fifth master category to which we can map our fifth category(Competency Level) then how will it impact SunbirdEd UI wherever Framework is getting rendered ? 
3- If there is no  fifth master category to which we can map our fifth category(Competency Level) then how can we create above framework with five categories in it ?

@vinukumar-vs @rajeevsathish @sharathprasad4tarento

**Comment by Krishnaj20** [comment]
tagging @pallakartheekreddy

**Comment by pallakartheekreddy** [comment]
@devendra-tarento 

We recommend you to create new master categories based on the required functionality. Below are the details to create new master category
https://knowlg.sunbird.org/learn/product-and-developer-guide/taxonomy-and-tagging/framework-service/faqs

> Do we have any other master category to which our fifth category(Competency Level) can be mapped ?
> If there is a fifth master category to which we can map our fifth category(Competency Level) then how will it impact SunbirdEd UI wherever Framework is getting rendered ?
> If there is no fifth master category to which we can map our fifth category(Competency Level) then how can we create above framework with five categories in it ?

Please try to use above recommendation. But, there is another category which is `topic` to use.

 @Krishnaj20

**Reply by devendra-tarento** [reply]
@pallakartheekreddy When we will add a new master category (Fifth category in our framework), In SunbirdEd UI wherever framework categories are getting rendered (currently with 4 categories as 4 dropdown) will start reflecting the fifth category with new dropdown if relevant form configuration updated ? 
Or We need to do code level changes in all those components where framework is getting rendered ?

**Reply by pallakartheekreddy** [reply]
@vinukumar-vs @rajeevsathish Can you check and update on this ?
@Krishnaj20

**Reply by vinukumar-vs** [reply]
Not in all places.. In some places it's configurable based on forms. You can updated the forms with new categories it will show the values.

Please try it & share if you are facing any issues.

@rajeevsathish @Krishnaj20

**Reply by devendra-tarento** [reply]
@vinukumar-vs Thanks, I will update on this.

**Reply by devendra-tarento** [reply]
@vinukumar-vs I have updated form configuration and able to get the new category and its values everywhere. Under workspace creator and reviewer are able to create, review and publish content with all five categories tagged to it successfully after modifying some schema in neo4j for QuestionSet. 

But on consumption screen we have below issues:

**1-** Framework filter with new category is visible but values in it are not coming. Hence we are not able to filter content based on new category.

**2-** Content are still being grouped by subject category assuming it is the last category but it is second last category in our case. We want them to be grouped by last category which is difficultyLevel (Competency Level).

 l looked at code and found that we are only expecting 4 categories and subject is always expected as last category.

Please suggest generic solution to address both issues. Attached screenshot for your reference:

<img width="1675" alt="Screenshot 2023-02-01 at 8 33 26 PM" src="https://user-images.githubusercontent.com/114464145/216082603-cd3278c8-abf8-4619-b95f-da1dafde1024.png">


@Krishnaj20 @rajeevsathish @pallakartheekreddy

**Reply by devendra-tarento** [reply]
@vinukumar-vs Could you please add your inputs asap so that we can make progress on it.

**Reply by vinukumar-vs** [reply]
We don't have this capability as of now. We are having this feature in release-5.2.0 ([JIRA story](https://project-sunbird.atlassian.net/browse/ED-306)) to add any property of the metadata to the filter list. 

> 1- Framework filter with new category is visible but values in it are not coming. Hence we are not able to filter content based on new category.


This should possible with the present Form configured for the menubar. 
> 2- Content are still being grouped by subject category assuming it is the last category but it is second last category in our case. We want them to be grouped by last category which is difficultyLevel (Competency Level).

@Krishnaj20 @rajeevsathish

**Reply by devendra-tarento** [reply]
> Is release-5.2.0 going to be next one after release-5.1.0 and Any date of its release ?
>> We don't have this capability as of now. We are having this feature in release-5.2.0 ([JIRA story](https://project-sunbird.atlassian.net/browse/ED-306)) to add any property of the metadata to the filter list.
>> 
>> > 1- Framework filter with new category is visible but values in it are not coming. Hence we are not able to filter content based on new category.
>> 
> Take a look at the course menu config in the menubar form config [here](https://jsonblob.com/1070976260511121408). ls there any thing wrong in this?
> 
>> This should possible with the present Form configured for the menubar.
>> 
>> > 2- Content are still being grouped by subject category assuming it is the last category but it is second last category in our case. We want them to be grouped by last category which is difficultyLevel (Competency Level).
>> 
> @Krishnaj20 @rajeevsathish

**Reply by devendra-tarento** [reply]
In content search api response for **Question Set**  getting correct facet values:

 **Request payload for Question set:**
<img width="826" alt="Content-search-QuestionSet-payload" src="https://user-images.githubusercontent.com/114464145/217259143-c8ef5242-47a8-4804-bc05-df1f1984219c.png">

 **Response for Question set:**
<img width="1013" alt="Content-search-questionset-response" src="https://user-images.githubusercontent.com/114464145/217259501-150255dc-1d0b-4742-9ba6-f1f94ba8c064.png">


But in content search api response for **Course** facet values are coming empty:

 **Request payload for Course:**
<img width="883" alt="Content-search-course-payload" src="https://user-images.githubusercontent.com/114464145/217260177-42b5ca23-e897-418d-ad09-27c4fe0bdfd0.png">


**Response for Course:**
<img width="951" alt="Content-search-course-response" src="https://user-images.githubusercontent.com/114464145/217260356-f9971c17-fa31-4eb1-b9d1-fa3a2aa039ec.png">


@vinukumar-vs @Krishnaj20 Please let me know if there is anything wrong here or any suggestion what could be wrong and how to fix it for Course

**Reply by vinukumar-vs** [reply]
@devendra-tarento can you verify difficultyLevel is part of Course schema or not? Can you share the Course Read API response?
@pallakartheekreddy @Krishnaj20

**Reply by devendra-tarento** [reply]
@vinukumar-vs We have update content-schema with difficultyLevel. But not getting difficultyLevel in the course read api response.

Is it because of difficultyLevel not being sent in course create/update api payload ?

Course update payload:
<img width="1680" alt="Course-update-payload" src="https://user-images.githubusercontent.com/114464145/218774338-f1a5bd15-c2f3-4f6a-a0c3-49cd4d4bd28f.png">

Now if I update the same course including board, medium, gradeLevel, subject, difficultyLevel in the payload via postman as below:
<img width="1085" alt="updated-course-update-payload" src="https://user-images.githubusercontent.com/114464145/218776871-dead25e1-6f07-46e5-a28f-7e74f9cec4a5.png">


After this course read api response returns the difficultyLevel in the response.

**Question:** How to pass board, medium, gradeLevel, subject, difficultyLevel in the payload when user create/update course on portal ? 
I tried to update course object category definition but it didn't work.

@pallakartheekreddy @Krishnaj20

**Reply by vinukumar-vs** [reply]
@devendra-tarento Can you verify the Schema file which is deployed as part of knowlg service for Course.
reference: 
https://github.com/project-sunbird/knowledge-platform/blob/release-5.1.0/schemas/collection/1.0/schema.json

@pallakartheekreddy @rajeevsathish

**Reply by devendra-tarento** [reply]
@vinukumar-vs Please find the Schema file which is deployed as part of knowlg service for Course below:
https://github.com/tsprasath/knowledge-platform/blob/Gcloud_fix/schemas/collection/1.0/schema.json

Both files are identical. Please let me know if anything missing in it.

@Krishnaj20 @pallakartheekreddy @rajeevsathish

**Comment by devendra-tarento** [comment]
@pallakartheekreddy I tried to add a new framework master category as mentioned below but getting error:
**Curl Request**

`curl --location --request POST 'https://uphrh.in//action/framework/v3/category/master/create' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer Token' \
--header 'x-authenticated-user-token: admin-token' \
--data-raw '{ 
    "request": { 
        "category":{ 
            "name":"Subject Difficulty Level", 
            "description":"Subject Difficulty Level", 
            "code":"difficultyLevel"
        } 
    } 
}'`

**Response**

{
    "id": "ekstep.learning.category.create",
    "ver": "1.0",
    "ts": "2023-01-22T17:11:20ZZ",
    "params": {
        "resmsgid": "4e333ff7-7dcf-49a2-8518-3a7bdf1d27c6",
        "msgid": null,
        "err": "ERR_GRAPH_ADD_NODE_VALIDATION_FAILED",
        "status": "failed",
        "errmsg": "Validation Errors"
    },
    "responseCode": "CLIENT_ERROR",
    "result": {
        "messages": [
            "Required Metadata orgIdFieldName not set",
            "Required Metadata targetIdFieldName not set",
            "Required Metadata searchIdFieldName not set",
            "Required Metadata searchLabelFieldName not set"
        ]
    }
}

Could you please suggest all steps which we need to take to get new framework master category added successfully ?

**Reply by pallakartheekreddy** [reply]
@devendra-tarento Please use below request in the creation of new master categories API
```
{
  "request": {
    "category": {
      "name": "Subject Difficulty Level",
      "description": "Subject Difficulty Level",
      "code": "difficultyLevel",
      "targetIdFieldName": "targetDifficultyLevelIds",
      "searchLabelFieldName": "se_difficultyLevels",
      "searchIdFieldName": "se_difficultyLevelIds",
      "orgIdFieldName": "difficultyLevelIds"
    }
  }
}
``` 
**Note:** We have fixed microsite documentation as well.
@Krishnaj20

**Reply by Krishnaj20** [reply]
@devendra-tarento - is your query clarified? or you have any follow up questions? Do let us know, accordingly we will lock this conversation.

**Reply by devendra-tarento** [reply]
@Krishnaj20 There are some open questions in above thread.

**Comment by Xitija** [comment]
Hello All,

As Karteekh Sir has pointed to the documentation which has a curl for creating a master category , on following that I tried the API through Postman. But I am facing the following error . 
`{
    "id": "api.error",
    "ver": "1.0",
    "ts": "2023-02-10 11:41:04:372+0000",
    "params": {
        "resmsgid": "ccdd8b40-a937-11ed-95ca-b17d34c345d0",
        "msgid": null,
        "status": "failed",
        "err": "FORBIDDEN_ERROR",
        "errmsg": "Forbidden: API WHITELIST Access is denied"
    },
    "responseCode": "FORBIDDEN",
    "result": {}
}`

**Reply by Krishnaj20** [reply]
@devendra-tarento - Since you were able to create the master category successfully,  can you respond to @Xitija 's query?

**Reply by Xitija** [reply]
Hello All,

Followed the steps given below to  create the master category

1. Check the kubernetes configmap for the content service to find which IP address is the learning service running on, and login to that VM via SSH
2. Check whether it is running with - `curl localhost:8080/learning-service/health`
3. If it is successful fire the command below (Add details regarding your category)
4. `curl --location --request POST 'http://localhost:8080/learning-service/framework/v3/category/master/create' 
--header 'Content-Type: application/json' 
--header 'Authorization: Bearer {{your key}}' 
--data-raw '{
    "request": {
        "category": {
            "name": "Contribution Type",
            "description": "Type of contribution",
            "code": "contributiontype",
            "targetIdFieldName": "targetContributionTypeIds",
            "searchLabelFieldName": "se_contributionType",
            "searchIdFieldName": "se_contributionTypeIds",
            "orgIdFieldName": "contributionTypeIds"
        }
    }
}'`
5. When the above API gives success response map the category by using the category create API (done through Postman)
6. `curl --location --request POST 'https://devnulp.niua.org/api/framework/v1/category/create?framework=tekdisandboxfw' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer {{your key}}' \
--data-raw '{
    "request": {
        "category": {
            "name": "Contribution Type",
            "description": "Type of contribution",
            "code": "contributiontype"
        }
    }
}'`
7. It gives a success response and mapped to the category we created on the backend

cc : @coolbung
